tests/std/tests/P1007R3_assume_aligned/test.cpp

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <assert.h>
+#include <memory>
+#include <stddef.h>
+
+void can_change_value_via_assume_aligned(int* p) {
+    int* p1 = std::assume_aligned<256>(p);
+    p1[0]   = 10;
+    p1[1]   = -15;
+}
+
+constexpr int can_read_constexpr_value_via_assume_aligned_and_sum_it(int const* p, size_t size) {
+    int const* p1 = std::assume_aligned<256>(p);
+    int result    = 0;
+    for (size_t i = 0; i < size; ++i) {
+        result += p1[i];
+    }
+    return result;
+}
+
+void assume_aligned_can_change_value() {
+    alignas(256) int arr[2] = {0, 0};
+    can_change_value_via_assume_aligned(&arr[0]);
+    assert(arr[0] == 10);
+    assert(arr[1] == -15);
+}
+
+constexpr void constexpr_assume_aligned_can_read_value() {
+    alignas(256) constexpr int arr[3] = {10, 11, 9};
+    static_assert(can_read_constexpr_value_via_assume_aligned_and_sum_it(arr, 3) == 30);
+}
+
+void assume_aligned_can_read_value() {
+    alignas(256) constexpr int arr[3] = {10, 11, 9};
+    assert(can_read_constexpr_value_via_assume_aligned_and_sum_it(arr, 3) == 30);
+}

```

Comment:
There's a technique to avoid repeating logic for runtime and `constexpr` test coverage. Here's what it looks like:

```
C:\Temp>type meow.cpp
```
```cpp
#include <assert.h>
#include <functional>
using namespace std;

constexpr bool test_multiplies() {
    multiplies<> function_object;
    assert(function_object(3, 4) == 12); // note: plain assert
    return true;
}

int main() {
    test_multiplies(); // test runtime behavior
    static_assert(test_multiplies()); // test compiletime behavior
}
```
```
C:\Temp>cl /EHsc /nologo /W4 /std:c++latest meow.cpp && meow
meow.cpp

C:\Temp>
```

That is, if you write a `constexpr` test function, it can contain ordinary `assert`s to verify things. If invoked normally at runtime, the `assert` statements will verify runtime behavior. If you `static_assert` the test function, that will force compile-time evaluation, and any failed `assert`s will be a compiler error. (This is not obvious - a successful assert does nothing, so it's constexpr-friendly. A failed assert will try to display a dialog and terminate the program, which is inherently constexpr-forbidden, so it will trigger a compiler error all by itself.)

I will push a simplification for this; I believe it can be extended to the modification test too.